onnxruntime_extensions/tools/pre_post_processing/steps/nlp.py

```diff
@@ -0,0 +1,508 @@
+
+        def build_output_imp():
+            output_str = []
+            for idx, out in enumerate(self.output_names):
+                output_str.append(f"{out} = Identity({self.input_names[idx % len(self.input_names)]})")
+
+            output_str = "\n".join(output_str)
+            return output_str
+
+        def build_input_declare():
+            inputs = []
+
+            for idx, inp in enumerate(self.input_names):
+                input_type_str_x, input_shape_str_x = self._get_input_type_and_shape_strs(graph, idx)
+                inputs.append(f"{input_type_str_x}[{input_shape_str_x}] {inp}")
+
+            return ",".join(inputs)
+
+        def build_tokenizer_call_arg():
+            call_args = f"{self.input_names[0]}"
+            return call_args
+
+        converter_graph = onnx.parser.parse_graph(
+            f"""\
+            qa_task ({build_input_declare()}) 
+                => ({build_output_declare()})
+            {{
+                {build_output_imp()}
+            }}
+            """
+        )
+
+        return converter_graph
+
+
+class BertTokenizerQATaskDecoder(Step):
+    def __init__(self, tokenizer_param: TokenizerParam, name: Optional[str] = None):
+        """
+        Brief:
+            Decode the input_ids to text
+        Args:
+            tokenizer_map: some essential infos, If you export tokenizer from hugging-face,
+            such as "tokenizer = transformers.AutoTokenizer.from_pretrained(model_name)"
+            You may need to provide the following:
+                tokenizer_param = TokenizerParam(vocab=tokenizer.vocab, #vocab is dict or file_path)
+            name: Optional name of step. Defaults to 'BertTokenizerQATaskDecoder'
+
+        """
+        # Bugs for default joins in onnxruntime_extensions/tools/pre_post_processing/pre_post_processor.py:187
+        # we have to put "input_ids_1" in the last to avoid failure connected
+        super().__init__(["start_logits", "end_logits", "input_ids_1"], ["text"], name)
+        self._domain = "com.microsoft.extensions"
+        self._tokenizer_param = tokenizer_param
+
+    def _create_graph_for_step(self, graph: onnx.GraphProto, onnx_opset: int):
+        graph_input_names = [inp.name for inp in graph.output]
+
+        # need to reorder input_names by similar to  "input_ids_1", "start_logits", "end_logits"
+        o_inputs = graph_input_names.copy()
+        for inp in o_inputs:
+            if "input_ids_1" in inp:
+                self.input_names[0] = inp
+            elif "start_logits" in inp:
+                self.input_names[1] = inp
+            else:
+                self.input_names[2] = inp
+        for inp in graph_input_names:
+            if self.input_names[2] in inp:
+                self.input_names[2] = inp
+        is_input_names_in_graph = [inp in graph_input_names for inp in self.input_names]
+        assert all(is_input_names_in_graph), f"Input names {self.input_names} not in graph inputs {graph_input_names}"
```

Comment:
Why does this get re-ordered inside the implementation vs. using an IOMapEntry to move inputs around if/when needed? 

Would have expected each step to simply define the inputs in the order it wants as we can't predict all the future use cases.

Reply:
Removed.